Gh 177

pana-cc opened the pull request:
WIP,
The mocha now uses core. There is still some work to do on the skipOnError,
Next step - adapt for jest, clean further mocha specific things.

pana-cc commented on packages/@testdeck/core/typings/index.d.ts:
```diff
@@ -163,7 +165,7 @@ export interface TestRunner<RunnerSuiteType, RunnerTestType> {
    * @param cb 
    */
-  declareTest(name: string, cb: (done?: Done) => void | Promise<void>);
+  declareTest(name: string, cb: MaybeAsyncCallback);
```
As discussed in the team channel... it is now :) I just pushed the latest changes.

pana-cc commented:
Check the changes again tomorrow, they should be much better, although not quite ready.

pana-cc commented:
A real unit test: https://github.com/testdeck/testdeck/blob/81cc2f15a13572c76b02a7b5b6ced2341cae4b5e/packages/%40testdeck/core/test.ts#L81
I am not sure this can hit 100% unit test coverage, but at least is a lot less messy than the integration tests we used to have.

Now that this won't ship as custom UI, and it won't require @types/mocha anymore it should be much more stable (less dependencies on typing, less branches to test) and I guess with high unit tests, integration tests can be simplified a lot.

pana-cc commented:
We can also make a pretty decent test runner ourselves. :D

silkentrance commented on packages/@testdeck/core/index.ts:
```diff
@@ -78,12 +78,12 @@ export abstract class ClassTestUI {
-    };
+    let settings = undefined;
+    if (ClassTestUI.slowSymbol in obj) (settings || (settings = {})).slow = obj[ClassTestUI.slowSymbol];
+    if (ClassTestUI.timeoutSymbol in obj) (settings || (settings = {})).slow = obj[ClassTestUI.timeoutSymbol];
```
.timeout?

silkentrance commented on packages/@testdeck/core/index.ts:
```diff
@@ -78,12 +78,12 @@ export abstract class ClassTestUI {
+    let settings = undefined;
+    if (ClassTestUI.slowSymbol in obj) (settings || (settings = {})).slow = obj[ClassTestUI.slowSymbol];
+    if (ClassTestUI.timeoutSymbol in obj) (settings || (settings = {})).slow = obj[ClassTestUI.timeoutSymbol];
+    if (ClassTestUI.retriesSymbol in obj) (settings || (settings = {})).slow = obj[ClassTestUI.retriesSymbol];
```
.retries?

silkentrance commented on packages/@testdeck/core/index.ts:
```diff
@@ -78,12 +78,12 @@ export abstract class ClassTestUI {
+    if (ClassTestUI.slowSymbol in obj) (settings || (settings = {})).slow = obj[ClassTestUI.slowSymbol];
+    if (ClassTestUI.timeoutSymbol in obj) (settings || (settings = {})).slow = obj[ClassTestUI.timeoutSymbol];
+    if (ClassTestUI.retriesSymbol in obj) (settings || (settings = {})).slow = obj[ClassTestUI.retriesSymbol];
+    if (ClassTestUI.executionSymbol in obj) (settings || (settings = {})).slow = obj[ClassTestUI.executionSymbol];
```
what is this for? skip, only?

silkentrance commented on packages/@testdeck/core/index.ts:
```diff
@@ -78,12 +78,12 @@ export abstract class ClassTestUI {
-      retries: obj[ClassTestUI.retriesSymbol],
-      execution: obj[ClassTestUI.executionSymbol]
-    };
+    let settings = undefined;
```
why do you not keep it as it was?

silkentrance commented:
> A real unit test:
[...]

Well, for this we would then be using whatever test framework and we will have to reimplement all the tests for all the supported test frameworks. Let it be copy & paste, since they all boil down to describe/it.

However, when fixing issues, adding additional tests, this will turn out to be a nightmare, maintenance wise.

Therefore my original approach with making this unit testable on the individual class that is assigned a very specific responsibility.

But let's see how this will turn out, shall we?

silkentrance commented:
Hm, from what I have gathered from the existing sources, you are currently trying to build a wrapper around the existing test framework, or are you not?

This might even break with jest which is always running tests in parallel.

Why not just keep it simple before trying to do such things in order to get it working?

silkentrance commented:
And maybe getting the test UI (aka user interface) thingy out of our minds will help. We just provide the user with an option to declare their test suites and tests in an object oriented way. Nothing more, and nothing less.

silkentrance commented:
Also, I do not like that we now have to use plain mocha describe/it for testing our own system. The system should provide for itself so that it can be tested against itself.

silkentrance commented:
Please keep it simple so we can have an initial release on npm. As I see it now, there is loads of work to be done to make it happen. Map and Reduce, please.

pana-cc commented:
I don't have time to explain myself right now, but I will do I promise, and I will split everything in smaller classes like in your branch. Also we will have less tests to maintain at the end.

silkentrance commented:
> I don't have time to explain myself right now, but I will do I promise, and I will split everything in smaller classes like in your branch. Also we will have less tests to maintain at the end.

Okay, I am looking forward to this. I just wondered, why we could not do the simple approach first, so that we have an early version of testdeck along with the various integrations for mocha and jest and possibly also jasmine, so that we can release it and have people use it, before doing the long haul.

pana-cc commented:
Because I catch bugs with a minimum set of unit tests at the moment. But the testrunner interface is small and reflected most of the feedback you had. If you want to kick in you can try to implement it for jest and jasmine, you can use the mocha as example. Don’t touch the core thoug as I am working on it and we ll have conflicts.

It has like 60-70 % coverage, there are two or three more important things that I want to cover before it can be used. After that splitting code in smaller classes will be a breeze since the unit tests now execute much faster.

Before there wer several implementations for things like ‘@test @timeout(100)’, ‘@test(timeout(100))’ etc. now these variations are covered with unit tests and generate the same “settings” for the runner this will greatly reduce the tests we have for actual runner implementations.

If I manage to clear few hours on this during the weekend I think it will get to a “beta” (without tests for the particular runners) ready to be published.

silkentrance commented:
@pana-cc What I meant was also that we do not require any tests on core right now. The tests in testdeck/mocha and others will suffice in order to get a working initial 0.0.1 release.

silkentrance commented:
@pana-cc I just took another look at the code, and in the test for core, the LoggingUI will mock away all calls to the underlying implementations.

While mocking is fine, this will only prove that the API works as expected, but not that the integration with jest/jasmine or mocha is working, or am I missing something?

silkentrance commented:
And while mocking the actual implementations of suite, test and whatnot in order to find out whether they are being called by testdeck/core, is actually a very good (!) idea, I think that these tests should be moved to the respective realizations of testdeck/core.

However, and since the test frameworks may change over time, we still need a full suite of integration tests there, to see whether everything is still working.

What do you think?

silkentrance commented:
I will work on a full suite of integration tests that we then reuse between the various realizations of testdeck/core, putting these into a private package under packages/@testdeck/test for example. We can then copy over the suite from there and integrate it to the specific packages during testing.

pana-cc commented:
You are right, mocha, jest and jasmine will need unit tests that will further validate that describe, it, xit, it.skip etc.

Keep in mind that now their tests will be reduced a lot. For example ‘params’ is completely handled in the core, for the actual runners the core will just call ‘it’ for every param.

pana-cc commented:
It sounds right, unit tests in each integration (mocha, jest, jasmine) that brings 100% unit test coverage for the integration package and a private test integration package with tests that can somehow be executed with every framework. I’ve been thinking to add the seeds as submodules for end-to-end tests. That will be a test pyramid. :)

silkentrance commented:
As far as I understood the future API, we would be using the `@testdeck/mocha` directly.

However, `@testdeck/mocha` will provide the core only with implementations of the required handlers for the decorators that are still provided by `@testdeck/core` or will it not?

pana-cc commented:
Public API, or how the people will use it, is they require @testdeck/mocha and use the ‘@suite’, ‘@test’ etc. it exports.

Under the hood the @testdeck/mocha provides with the implementation of the minimal glue code and reexports all from @testdeck/core

Since @testdeck/mocha has like 100 lines of code it should be possible to unit test it with very few tests.

And to assure it is all piped properly there should be some end-to-end tests. But these should be very very few. Since they should just ensure the interface implementations are passed around and reexported properly.

silkentrance commented:
So we are on the same page, considering the following "analysis"?

![image](https://user-images.githubusercontent.com/6068824/54558593-6c3c0f00-49be-11e9-9b34-b2a567f27d24.png)

silkentrance commented:
Of course, the lot of ParametrisedTestDeclaration, TestDeclaration and SuiteDeclaration need some more work as we do not want user to do `@suite.skip.only.skip.pending.pending...`

pana-cc commented:
The public API is fine.
The standard API - not so much, I am not sure what the TestdeckUI and RegisteredUI is. There is no registration, the mocha/jest/jasmine will simply implement the TestRunner interface and pass it to the core (right now through the ClassTestUI constructor): https://github.com/testdeck/testdeck/blob/bba9101b8473430fbdd0d9d565ce8f499f4e28b6/packages/%40testdeck/mocha/index.ts#L73

silkentrance commented:
> The public API is fine.
> The standard API - not so much, I am not sure what the TestdeckUI and RegisteredUI is. There is no registration, the mocha/jest/jasmine will simply implement the TestRunner interface and pass it to the core (right now through the ClassTestUI constructor):
> 
> [testdeck/packages/@testdeck/mocha/index.ts](https://github.com/testdeck/testdeck/blob/bba9101b8473430fbdd0d9d565ce8f499f4e28b6/packages/%40testdeck/mocha/index.ts#L73)
> 
> Line 73 in [bba9101](/testdeck/testdeck/commit/bba9101b8473430fbdd0d9d565ce8f499f4e28b6)
> 
>  public constructor(runner: core.TestRunner = mochaRunner) {

What I am missing in the "standard" testdeck UI is the decorators. The test runner, as your defined it, will not make the trick as it does not decorate the class or method.

As such, I think of it as a registered UI that will be called by our "framework" in order to declare the suites and tests, with the rest being done by the actual "framework' which is testdeck/core.

pana-cc commented:
Well, it is fine then :)
I hope I will be able to add DI tests and tests for throwing callbacks today, the core will be all set for unit tests.
Then we can publish v0.0.1

pana-cc commented:
If the developers will require `@testdeck/mocha` the test's won't be portable, e.g. once you write tests for the mocha runner you will have to replace the imports if you want to run them with jest or jasmine. Maybe core should really export the decorators too and @testdeck mocha/jest/jasmine should register handlers for them.

pana-cc commented:
I've added few unit tests about jest and implemented most of the basic stuff. Timeouts are not set yet, if skip is not supported I guess we can just skip it. I've published the packages to npm under the @testdeck org. The uploaded files are probably wrong but I ll check them over the weekend.

pana-cc commented:
Jest, jasmine and mocha are now supported and there are seeds working for the three of them. I am going to merge gh-177 soon into master. There are READMEs, some integrations tests, and a few other minor things to do but it is close to ready.

silkentrance commented:
I have just looked at the releases over at NPM. We will have to work on the documentation part, as the examples are referring to packages that do not exist, e.g. testdeck-jest instead of `@testdeck/jest` and so on.

pana-cc commented:
We can probably make a `@testdeck/testdeck` that gets its implementation from one of the `@testdeck/jest`, `@testdeck/mocha` or `@testdeck/jasmine`. When tests are written to use `@testdeck/testdeck` they could be completely portable. However I am not sure how this will work with web pack etc. Mocha has this option to specify packages to be required at start, so there a `--require @testdeck/mocha` in the mocha.opts can load the mocha integration that will enable `@testdeck/testdeck` decorators to work with mocha... If there is something similar for the other frameworks maybe the tests can be protable without even having to replace the integration import in each test file.

silkentrance commented:
In order to get the failing tests working I have applied the following patch

```
diff --git a/packages/@testdeck/core/test.ts b/packages/@testdeck/core/test.ts
index 18c8516..272c9d4 100644
--- a/packages/@testdeck/core/test.ts
+++ b/packages/@testdeck/core/test.ts
@@ -111,6 +111,11 @@ namespace LoggingClassTestUI {
     }
 }

+function cleancov(s: string): string {
+
+    return s.replace(/cov_[^.]+(f|s)[^+]+[+][+];/g, "");
+}
+
 describe("testdeck", function() {
     let ui: LoggingClassTestUI;
     beforeEach("create ui", function() {
@@ -520,7 +525,7 @@ describe("testdeck", function() {
             assert.equal(callbacks[0].toString(), 'before() { cycle.push("Before All"); }');

             assert.equal(callbacks[1].name, "setupInstance");
-            assert.equal(callbacks[1].toString(), "function setupInstance(){cov_r11ocdey9.f[14]++;cov_r11ocdey9.s[43]++;instance=theTestUI.createInstance(constructor);}");
+            assert.equal(cleancov(callbacks[1].toString()), "function setupInstance(){instance=theTestUI.createInstance(constructor);}");

             assert.equal(callbacks[2].name, "before");
             assert.equal(callbacks[2].toString(), 'before() { cycle.push("Before Each"); }');
@@ -532,7 +537,7 @@ describe("testdeck", function() {
             assert.equal(callbacks[4].toString(), 'after() { cycle.push("After Each"); }');

             assert.equal(callbacks[5].name, "teardownInstance");
-            assert.equal(callbacks[5].toString(), "function teardownInstance(){cov_r11ocdey9.f[27]++;cov_r11ocdey9.s[94]++;instance=null;}");
+            assert.equal(cleancov(callbacks[5].toString()), "function teardownInstance(){instance=null;}");

             assert.equal(callbacks[6].name, "after");
             assert.equal(callbacks[6].toString(), 'after() { cycle.push("After All"); }');
```

pana-cc commented:
I am getting out of juice. Next month I'll be traveling for Microsoft Build and I'll have to prepare for it. May I merge this PR into master? There are a lot of stuff to do but jest, jasmine and mocha somewhat work now. After this is merged we can write up the work items left, triage and prioritize them.

silkentrance commented:
Okay, this was definitely a sprint 👍

silkentrance commented:
I can work out the remaining issues and also work on the documentation.

silkentrance commented:
Perhaps getting this patch into the tests so that the tests will not fail? https://github.com/testdeck/testdeck/pull/181#issuecomment-476387308

pana-cc commented:
This test failed on my end too. The nyc maybe inserts code to track coverage and it changes over time, but leaks through the function's toString. Asserting on callback[5].toString is no good in this state, maybe just remove that assert?

silkentrance commented:
Don't we want to make sure that the new instance was created and torn down?

pana-cc commented:
Go ahead and either merge the patch or even remove the asserts on the callbacks that contain `cov_r11ocdey9` traces.

silkentrance commented:
Alternatively I can make the commit so that the tests will run.

pana-cc commented:
This test asserts the "wrap" function properly transitions the original function's name and toString. This was to display proper function bodies in somewhat browser based test runner, I can't remember exactly. There are other tests that will actually execute the callbacks and assert the instance is created.

silkentrance commented:
Do not pull the changes yet. I will still have to make some changes.

pana-cc commented:
Ok make the commit :) and merge this all thing into master.

TODO:
 - [ ] Integration tests are ignored at this point, we need to pick 1-2 tests with "only" and 1-2 of the most extensive suites. I think with the new unit tests in place the old way of testing 100% coverage with end-to-end tests is a bit excessive. They used to take long. Maybe now we can take a smaller subset that will execute faster. Also the integration tests does not need to pack and install, we can rely on lerna to symlink what is necessary into the integration tests.
 - [ ] Lerna publish failed on private packages (the seeds/*) maybe there is a way to list just the @testdeck/* as publishable
 - [ ] Readme for each: jest, mocha, jasmine
 - [ ] Testdeck domain? Publish the docs.
 - [ ] Update the docs.
 - [ ] skipOnError
 - [ ] retries
 - [ ] Continue the refactoring, split the big class in core into smaller classes (just keep them in the same file, I kind of get lost when the code is moved into smaller files, idk I am used to folding code) Overall this should be relatively safe since there is a decent unit test coverage in core, the tests there run for few seconds, so iterations should be fast.
 - [ ] Notify all contributors about all that happened, maybe contributors will be more engage and give it a spin, catch bugs etc.
 - [ ] Try to add the testdeck interface in the wikis/readmes to jest, jasmine, mocha.
 - [ ] Write a blog (in GitHub pages) about the mocha/jest jasmine, post it in some social groups, reddit, etc. make some noise

i am probably missing something, :( there so much more work to be done here.

pana-cc commented:
- [ ] Watcher? Move it into separate package, remove it from @testdeck/mocha. Make it configurable to run mocha/jest/jasmine after tsc change detection.
 - [ ] Watcher tests.

pana-cc commented:
- [ ] Merge the logo, use it in the READMEs

silkentrance commented:
In the top level package there is a dev dependency tsc, which is deprecated, I think that we can remove it as the typescript dependency will provide us with the tsc command.

silkentrance commented:
Builds are working on appveyor, however, the regex fails to match on travis-ci, huh?

silkentrance commented:
I will need some extra turns on a branch of this in order to get things working on travis-ci, gh-177-3 ftw.

silkentrance commented:
Makes you wonder why this works on OSX and Windows but not on linux :)

```
    return s.replace(/cov_[^.]+[.](f|s)[^+]+[+][+];/g, "");
                                                 ^-- forgot this one here
```

silkentrance commented:
As for the private package thing

https://github.com/lerna/lerna/pull/846

this should have been resolved by now.

silkentrance commented:
@pana-cc the seed/* folders are empty. Is this by design?

silkentrance commented:
Just saw that the seeds are pulled in as sub modules, well this requires additional changes to travis/appveyor.

silkentrance commented:
Builds are now working on both travis and appveyor, lest the pending conflicts with master.

I wonder, though, why we also package the tests for distribution. Should we not keep the distribution as minimal as possible?

silkentrance commented:
Rebased to master

silkentrance commented:
Okay, the tests are not part of the distribution. Just realised this.

pana-cc commented:
For the ‘seeds/*’ use ‘git submodule update --init’. Submodules are like checked out repos with specific commit. If you make changes inside then you have to cd to seeds/testdeck-mocha-seed there make a commit and push, then go to the root and git add the seeds/test deck-mocha-seed, commit (this will update the commit sha of the submodule) and push testdeck. Look for “git submodules”.

pana-cc commented:
That RegExp, I love your pragmatism. :)
